src/gql/schema.rs

```diff
@@ -0,0 +1,626 @@
+
+		let mut table_ty_obj = Object::new(tb.name.to_string())
+			.field(Field::new("id", TypeRef::named_nn(TypeRef::ID), |ctx| {
+				FieldFuture::new(async move {
+					let record = ctx.parent_value.as_value().unwrap();
+					let GqlValue::Object(record_map) = record else {
+						todo!()
+					};
+					let id = record_map.get("id").unwrap();
+
+					Ok(Some(id.to_owned()))
+				})
+			}))
+			.implement(interface);
+
+		for fd in fds.iter() {
+			let fd_name = Name::new(fd.name.to_string());
+			let fd_type = kind_to_type(fd.kind.clone());
+			table_orderable = table_orderable.item(fd_name.to_string());
+			let type_filter_name = format!("_filter_{}", unwrap_type(fd_type.clone()));
+
+			let type_filter =
+				Type::InputObject(filter_from_type(fd.kind.clone(), type_filter_name.clone()));
+			trace!("\n{type_filter:?}\n");
+			types.push(type_filter);
+
+			table_filter = table_filter
+				.field(InputValue::new(fd.name.to_string(), TypeRef::named(type_filter_name)));
+
+			table_ty_obj =
+				table_ty_obj.field(Field::new(fd.name.to_string(), fd_type, move |ctx| {
+					let fd_name = fd_name.clone();
+					FieldFuture::new(async move {
+						let record = ctx.parent_value.as_value().unwrap();
+						let GqlValue::Object(record_map) = record else {
+							todo!("got unexpected: {record:?}, processing field {fd_name}")
+						};
+						let val = record_map.get(&fd_name).unwrap();
+
+						Ok(Some(val.to_owned()))
+					})
+				}));
+		}
+
+		types.push(Type::Object(table_ty_obj));
+		types.push(table_order.into());
+		types.push(Type::Enum(table_orderable));
+		trace!("\n\n\n{table_filter:?}");
+		types.push(Type::InputObject(table_filter));
+	}
+
+	trace!("current Query object for schema: {:?}", query);
+
+	let mut schema = Schema::build("Query", None, None).register(query);
+	for ty in types {
+		trace!("adding type: {ty:?}");
+		schema = schema.register(ty);
+	}
+
+	let id_interface =
+		Interface::new("record").field(InterfaceField::new("id", TypeRef::named_nn(TypeRef::ID)));
+	schema = schema.register(id_interface);
+
+	// TODO: when used get: `Result::unwrap()` on an `Err` value: SchemaError("Field \"like.in\" is not sub-type of \"relation.in\"")
+	let relation_interface = Interface::new("relation")
+		.field(InterfaceField::new("id", TypeRef::named_nn(TypeRef::ID)))
+		.field(InterfaceField::new("in", TypeRef::named_nn("record")))
+		.field(InterfaceField::new("out", TypeRef::named_nn("record")))
+		.implement("record");
+	schema = schema.register(relation_interface);
+
+	let res = schema.finish();
+	match res {
+		Ok(s) => Ok(s),
+		Err(_) => {
+			warn!("note: Query reloading is not yet implemented. This means that the schema is generated on startup, so if starting in memory the schema will always be empty");
+			res.unwrap();
+			unreachable!();
+		}
+	}
+
+	// Ok(schema.finish().unwrap())
+}
+
+fn sql_value_to_gql_value(v: SqlValue) -> Result<GqlValue, ()> {
+	let out = match v {
+		SqlValue::None => GqlValue::Null,
```

Comment:
I know that `null` is considered the `undefined`/`none` of graphql, but in the case of SurrealQL they are two different values, used for example when removing fields from a record's document. Should we introduce a special `None` placeholder type?

Reply:
On the GQL side, when consumed (mostly by JS / TS clients), this usually refers to just `null`, having another placeholder type makes consumption of the GQL API more verbose IMO. Personally, I'd have the currently implementation to keep things simpler and to GraphQL's specification.